### ISSUE TITLE
Type inference into function

### ISSUE THREAD

**Issue by artificial-aidan** [python/mypy#14236]
**Feature**

I tried to search using some keywords and couldn't find anything obvious. But I was trying to do type validation of a class that I don't control (external sdk) and it has a class with optional fields. I'd like to write a function to validate that the instance I am passed as the correct field types, instead of doing it all in one.

**Pitch**

Ability to do type validation of objects you don't have control of in a subfunction, instead of in a main function.

```
import typing as t


class Test:
    opt: t.Optional[str]


def validate_test(t: Test) -> None:
    assert (t.opt)


def check_test_not_working() -> None:
    t = Test()
    validate_test(t)
    # Expect str, get Union[str, None]
    reveal_type(t.opt)


def check_test_working() -> None:
    t = Test()
    assert (t.opt)
    # Expect str, get str
    reveal_type(t.opt)
```

**Comment by erictraut** [comment]
When you say "type validation", are you referring to runtime validation (as in `assert t.opt`), or are you referring to static type validation (ensuring that the type checker evaluates an expected type of a particular expression). If it's the latter, the recently-introduced `typing.assert_type` is probably what you're looking for. It's also available in `typing_extensions` for versions of Python prior to 3.11.

```python
from typing_extensions import assert_type

def check_test_not_working() -> None:
    t = Test()
    validate_test(t)
    assert_type(t.opt, str | None)

def check_test_working() -> None:
    t = Test()
    assert t.opt
    assert_type(t.opt, str)
```

**Comment by artificial-aidan** [comment]
Sorry, I was just using assert as an example of type narrowing. Currently when an assert is used the type checker can narrow the expected type to not include `None`. What I'm suggesting is doing that type narrowing by following functions as well.

```
import typing as t


class Test:
    opt: t.Optional[str]


def validate_test(t: Test) -> None:
    assert (t.opt)


def check_test_not_working() -> None:
    t = Test()
    validate_test(t)
    # Expect str, get Union[str, None]
    reveal_type(t.opt)


def check_test_working() -> None:
    t = Test()
    if !t.opt:
        # return here if None, so below the type is narrowed to str
        return
    # Expect str, get str
    reveal_type(t.opt)
```

**Comment by hauntsaninja** [comment]
There is no way to express the post condition you want in the current type system. Assuming you don't want to refactor your code, I would just add an extra `assert t.opt is not None` after the return, which will make things work.

In some similar cases, TypeGuard can be made to work: https://peps.python.org/pep-0647/ 
I think use cases like yours were briefly discussed when PEP 647 was being drafted, Eric might remember better.

**Comment by artificial-aidan** [comment]
I couldn't figure out how to make TypeGuard work with a class that I don't maintain.

What do you mean after the return?

My goal (which I accept may not be possible) is to perform all the type narrowing in a separate function, instead of doing it inline with the rest of the code.

**Comment by A5rocks** [comment]
Not *exactly* your case (as @hauntsaninja mentioned, TypeGuard doesn't work for your case), but as an example of TypeGuard:

```py
import typing as t

class Test:
    opt: t.Optional[str]

class GoodTest(Test):
    opt: str


def validate_test(t: Test) -> t.TypeGuard[GoodTest]:
    return bool(t.opt)


def check_test_not_working() -> None:
    t = Test()
    assert validate_test(t)
    # Expect str, get Union[str, None]
    reveal_type(t.opt)


def check_test_working() -> None:
    t = Test()
    if not t.opt:
        # return here if None, so below the type is narrowed to str
        return
    # Expect str, get str
    reveal_type(t.opt)
```

Essentially, just make a subclass of the foreign class.

**Comment by artificial-aidan** [comment]
Yeah, subclass is where I ended up. But for the sake of maintainability I decided not to.

This would be a nice feature, and I wanted to make sure it wasn't already supported. So I think my feature request still stands if it's understood.
